orchagent/vnetorch.cpp

```diff
@@ -2004,6 +2014,160 @@ bool VNetRouteOrch::delOperation(const Request& request)
+                SWSS_LOG_INFO("Prefix %s covers destination address",
+                    route.first.to_string().c_str());
+                observerEntry->second.routeTable.emplace(
+                    route.first,
+                    route.second
+                );
+            }
+        }
+    }
+
+    observerEntry->second.observers.push_back(observer);
+
+    SWSS_LOG_NOTICE("Attached next hop observer of route %s for destination IP %s",
+        observerEntry->second.routeTable.rbegin()->first.to_string().c_str(),
+        dstAddr.to_string().c_str());
+
+    for (const auto & route : observerEntry->second.routeTable)
+    {
+        VNetNextHopUpdate update = {
+            SET_COMMAND,
+            std::get<0>(route.second), // vnet name
+            dstAddr, // destination
+            route.first, // prefix
+            std::get<1>(route.second) // nexthop
+        };
+        observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
+    }
+}
+
+void VNetRouteOrch::detach(Observer* observer, const IpAddress& dstAddr)
+{
+    SWSS_LOG_ENTER();
+    auto observerEntry = next_hop_observers_.find(dstAddr);
+
+    if (observerEntry == next_hop_observers_.end())
+    {
+        SWSS_LOG_ERROR("Failed to detach observer for %s. Entry not found.", dstAddr.to_string().c_str());
+        assert(false);
+        return;
+    }
+    auto iter = std::find(
+        observerEntry->second.observers.begin(),
+        observerEntry->second.observers.end(),
+        observer);
+    if (iter == observerEntry->second.observers.end())
+    {
+        SWSS_LOG_ERROR("Failed to detach observer for %s. Observer not found.", dstAddr.to_string().c_str());
+        assert(false);
+        return;
+    }
+    for (const auto& route : observerEntry->second.routeTable)
+    {
+        VNetNextHopUpdate update = {
+            DEL_COMMAND,
+            std::get<0>(route.second), // vnet name
+            dstAddr, // destination
+            route.first, // prefix
+            std::get<1>(route.second) // nexthop
+        };
+        observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
+    }
+    next_hop_observers_.erase(observerEntry);
+}
+
+void VNetRouteOrch::addRoute(const std::string& vnet, const IpPrefix& ipPrefix, const nextHop& nh)
+{
+    SWSS_LOG_ENTER();
+
+    for (auto & next_hop_observer : next_hop_observers_)
+    {
+        if (ipPrefix.isAddressInSubnet(next_hop_observer.first))
+        {
+            next_hop_observer.second.routeTable.emplace(ipPrefix, std::make_tuple(vnet, nh));
+            for (auto& observer : next_hop_observer.second.observers)
+            {
+                VNetNextHopUpdate update = { 
+                    SET_COMMAND, 
+                    vnet,
+                    next_hop_observer.first, 
+                    ipPrefix, 
+                    nh
+                };
+                observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
```

Comment:
do you need to check if the newly added route is a 'better' match for the current observers?

Reply:
Let's discuss it at teams.

Reply:
Done~ Please review the newest.